.github/reusable-audit.yml

```diff
@@ -0,0 +1,46 @@
+name: Reusable Audit Workflow
+
+on:
+  workflow_call:
+    inputs:
+      useForceResolutions:
+        required: false
+        default: false
+        type: boolean
+    secrets:
+      WEBHOOK_E2E_FAILURE:
+        required: true
+
+jobs:
+  audit:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2.4.0
+      - uses: actions/setup-node@v2.5.1
+        with:
+          node-version: "16.x"
+          registry-url: "https://registry.npmjs.org"
+      - run: npm ci
+      # We check the dependency licenses as an audit task:
+      - run: npx license-checker --production --failOn "AGPL-1.0-only; AGPL-1.0-or-later; AGPL-3.0-only; AGPL-3.0-or-later; Beerware; CC-BY-NC-1.0; CC-BY-NC-2.0; CC-BY-NC-2.5; CC-BY-NC-3.0; CC-BY-NC-4.0; CC-BY-NC-ND-1.0; CC-BY-NC-ND-2.0; CC-BY-NC-ND-2.5; CC-BY-NC-ND-3.0; CC-BY-NC-ND-4.0; CC-BY-NC-SA-1.0; CC-BY-NC-SA-2.0; CC-BY-NC-SA-2.5; CC-BY-NC-SA-3.0; CC-BY-NC-SA-4.0; CPAL-1.0; EUPL-1.0; EUPL-1.1; EUPL-1.1;  GPL-1.0-only; GPL-1.0-or-later; GPL-2.0-only;  GPL-2.0-or-later; GPL-3.0; GPL-3.0-only; GPL-3.0-or-later; SISSL;  SISSL-1.2; WTFPL"
+      # Some issues identified by npm audit cannot be fixed by us
+      # (i.e. they require an upstream dependency update),
+      # but using `npm-force-resolutions` we can at least override the upstream
+      # dependency when running an audit.
+      #
+      # If this task fails, first attempt to run `npm audit fix` or upgrade the
+      # relevant dependency; when that does not work, you can add the transitive
+      # dependency to the `resolutions` field in package.json according to the
+      # instructions in `npm-force-resolutions`'s README.
+      - name: Setup npm-force-resolutions
+        run: npx npm-force-resolutions
+        if: ${{ inputs.useForceResolutions }}
```

Comment:
I don't think that's what we want.

We want to see the task fail and be notified and update our packages accordingly.

Reply:
This currently won't generally be used, it's really there for "this was currently needed in one place, so having it there as an option can't hurt until this is stable"

Reply:
Auditing is simple:
1. Check the package lock, fail if anything is vulnerable there

We don't want any external package even involved in this workflow.

There is a reason for audit to be a built in npm command. For example, what if npm-force-resolution is the corrupt package?

Nothing extraneous should interfere with audit.

1. checkout
2. run audit

By the way, I don't know if you've noticed, but there is still an `npm ci` step here, which defeats the purpose of an audit. You run the audit before you install a vulnerable package.

Also, we agreed as a team to use major versions for github actions.

Reply:
The only package that currently does what you describe is solid-client-js, which only recently had that refactored — for our other repos (e.g., [solid-client-authn-js](https://github.com/inrupt/solid-client-authn-js/blob/3cdc8595d90491aa3166893150e0e24c6dfadb0c/.github/workflows/audit.yml#L21), the still use as currently here — we can improve later, this is a first pass: move things to a central place, then we can refactor all together.

As for anything run by `npx`, it's entirely outside of the audit pipeline, further, this is gated by an `if` — ideally we can remove this in 1 or 2 iterations.

Reply:
We've had cases in the past where, several layers deep in a dev dependency that had no impact on anything, that we've wanted to ignore so that we would be alerted when more high-priority packages have vulnerabilities. This is a way to cut down on the noise so we don't get in the habit of ignoring a big red X. It's still absolutely a hack, but there are tradeoffs; get used to seeing failures, or remember to have to audit the force-resolutions list occasionally.

This works as a first pass, and we should move forward with removing force-resolutions where we can.

Reply:
The purpose of running `npm-force-resolutions` is to be able to update transitive dependencies in the case they have a known vulnerability, but the direct dependency pulling it in hasn't been updated yet (see their [readme](https://www.npmjs.com/package/npm-force-resolutions) ). Whether we want to systematically run it as described in their readme or to control whether it should run or not as it is done here is something we can discuss, but if do use this tool it has to run as part of the audit CI pipeline, otherwise I think it defeats the purpose. Which means I don't think simply checking out and running audit is something we could do if we wanted to use this, which I think we do (it proved helpful a couple of times already).

Reply:
OK, well then in any case, let's not make it a default to hide audit failures.

Reply:
Also, let me ask a question:

I checked out npm-force-resolutions a bit and would it be REALLY stupid when there is a lingering messed up transitive dependency to push a patched up package-lock.json to main instead of relying on dodgy preinstall hooks?

Reply:
I had the same question, what is the reason not to use `npm-force-resolutions` manually and commit the result ?

Reply:
I think it's really not worth us going further into this — npm-force-resolutions is/was used on one package, and it made sense to support that whilst we rework the CI/CD in general to be more standardised — it's okay that this isn't the ultimate standard just yet, we can incrementally change things

Reply:
As for dev vs production dependencies, we probably want a "warning" for devDependencies, but an "error" for dependencies — `npm audit` can do that via `--only=dev` or `--only=prod`: https://docs.npmjs.com/cli/v8/commands/npm-audit